Simplify CUDAMultiStreamGuard

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57052 Merge CUDAFuture into ivalue::Future
* #57051 Make CUDAFuture handle any kind of device type
* #57050 Use only generic helpers in CUDAFuture
* #57049 Introduce generic MultiStreamGuard
* **#57048 Simplify CUDAMultiStreamGuard**
* #57047 Add recordDataPtrOnStream to DeviceGuardImplInterface
* #57046 Add getStreamFromPool to DeviceGuardImplInterface

CUDAMultiStreamGuard had a default constructor and a `original_devices()` method which were only used in a test. I'm removing them here to simplify the API and make it easier to manipulate this class later. One extra benefit is that this class used to get and store the current stream of _all_ devices, whereas now it only does so for the relevant devices.

Differential Revision: [D28029160](https://our.internmc.facebook.com/intern/diff/D28029160/)

lw committed:
Simplify CUDAMultiStreamGuard

CUDAMultiStreamGuard had a default constructor and a `original_devices()` method which were only used in a test. I'm removing them here to simplify the API and make it easier to manipulate this class later. One extra benefit is that this class used to get and store the current stream of _all_ devices, whereas now it only does so for the relevant devices.

Differential Revision: [D28029160](https://our.internmc.facebook.com/intern/diff/D28029160/)

[ghstack-poisoned]

lw committed:
Update on "Simplify CUDAMultiStreamGuard"


CUDAMultiStreamGuard had a default constructor and a `original_devices()` method which were only used in a test. I'm removing them here to simplify the API and make it easier to manipulate this class later. One extra benefit is that this class used to get and store the current stream of _all_ devices, whereas now it only does so for the relevant devices.

Differential Revision: [D28029160](https://our.internmc.facebook.com/intern/diff/D28029160/)

[ghstack-poisoned]

lw committed:
Update on "Simplify CUDAMultiStreamGuard"


CUDAMultiStreamGuard had a default constructor and a `original_devices()` method which were only used in a test. I'm removing them here to simplify the API and make it easier to manipulate this class later. One extra benefit is that this class used to get and store the current stream of _all_ devices, whereas now it only does so for the relevant devices.

Differential Revision: [D28029160](https://our.internmc.facebook.com/intern/diff/D28029160/)

[ghstack-poisoned]

mrshenli commented on aten/src/ATen/cuda/CUDAMultiStreamGuard.h:
```diff
@@ -16,32 +16,22 @@ class CUDAMultiStreamGuard final {
-    }
-  }
-
-  CUDAMultiStreamGuard(const CUDAGuard&) = delete;
```
curious, why existing deleted assignment and move constructors took a `CUDAGuard` instead of a `CUDAMultiStreamGuard`?

lw replied:
I suspect this came from a botched copy-paste.

mrshenli commented on aten/src/ATen/test/cuda_stream_test.cpp:
```diff
@@ -145,14 +144,6 @@ TEST(TestStream, CUDAGuardTest) {
 
-  // Test that all original streams are recorded.
-  {
-    at::cuda::CUDAMultiStreamGuard guard;
```
Are there any other tests covering `CUDAMultiStreamGuard`? If not, shall we add one?

lw replied:
I grepped through the code and found only this one. I can see if I find a good place to add a few new ones.

lw committed:
Update on "Simplify CUDAMultiStreamGuard"


CUDAMultiStreamGuard had a default constructor and a `original_devices()` method which were only used in a test. I'm removing them here to simplify the API and make it easier to manipulate this class later. One extra benefit is that this class used to get and store the current stream of _all_ devices, whereas now it only does so for the relevant devices.

Differential Revision: [D28029160](https://our.internmc.facebook.com/intern/diff/D28029160/)

[ghstack-poisoned]
